// $actions/layoutActions.js
import {
  LAYOUT_PANEL_SET_ITEMS_STATE,
  LAYOUT_PANEL_PUSH_ITEMS_STATE,
  LAYOUT_PANEL_POP_ITEMS_STATE,

  LAYOUT_PANEL_PUSH_MESSAGE,
  LAYOUT_PANEL_REMOVE_MESSAGE,

  LAYOUT_CLOSE_MODAL,
  LAYOUT_CLOSE_MODALS,
  LAYOUT_OPEN_MODAL
} from '$actionTypes';

export const openModal = (modal) => {
  return {
    type: LAYOUT_OPEN_MODAL,
    payload: modal
  }
}

export const closeModal = () => {
  return {
    type: LAYOUT_CLOSE_MODAL
  }
}

export const closeModals = () => {
  return {
    type: LAYOUT_CLOSE_MODALS
  }
}

export const setNavPanelItemsState = (items) => {
  return {
    type: LAYOUT_PANEL_SET_ITEMS_STATE,
    payload: items
  }
}

export const pushNavPanelItemsState = (item) => {
  return {
    type: LAYOUT_PANEL_PUSH_ITEMS_STATE,
    payload: item
  }
}


export const pushMessage = mes => dispatch => {
  let message = {
    text: mes,
    id: Date.now()
  };
  
  setTimeout(() => dispatch({ type: LAYOUT_PANEL_REMOVE_MESSAGE }), 2500);

  dispatch({
    type: LAYOUT_PANEL_PUSH_MESSAGE,
    payload: message
  })
}


export const popNavPanelItemsState = (item) => {
  return {
    type: LAYOUT_PANEL_POP_ITEMS_STATE
  }
}
