// $actions/phonesPageActions.js
import {
  FETCH_PRODUCTS_START,
  FETCH_PRODUCTS_FAIL,
  FETCH_PRODUCTS_SUCCESS,
  PRODUCTS_IMAGE_LOADED,
  PRODUCTS_SAVE_SCROLL_POS,
  PRODUCTS_FILTER,

  CLEAN_PRODUCTS
  
  /* LAYOUT_PANEL_LOADING_TRUE,
  LAYOUT_PANEL_LOADING_FALSE */

} from '$actionTypes'

import { fetchPhones as fetchPhonesApi } from '$api/index';

export const fetchPhones = () => async (dispatch, getState) => {
  dispatch({ type: FETCH_PRODUCTS_START })

  try {
    let phones = await fetchPhonesApi(getState().phonesPage.ids.length);
    console.log('SKIPPED ' + getState().phonesPage.ids.length);
    dispatch({
      type: FETCH_PRODUCTS_SUCCESS,
      payload: phones
    })
  } catch (error) {
    dispatch({
      type: FETCH_PRODUCTS_FAIL,
      payload: error,
      error: true
    })
  }
}

export const cleanProducts = () => {
  return {
    type: CLEAN_PRODUCTS
  }
}

export const onImageLoaded = (item) => {
  return {
    type: PRODUCTS_IMAGE_LOADED,
    payload: item
  }
}

export const saveScrollPos = (pos) => {
  return {
    type: PRODUCTS_SAVE_SCROLL_POS,
    payload: pos
  }
}

export const filter = (query) => {
  return {
    type: PRODUCTS_FILTER,
    payload: query
  }
}
