// $reducers/layout.js
import {

  LAYOUT_PANEL_LOADING_TRUE, 
  LAYOUT_PANEL_LOADING_FALSE,

  LAYOUT_PANEL_SET_ITEMS_STATE,
  LAYOUT_PANEL_PUSH_ITEMS_STATE,
  LAYOUT_PANEL_POP_ITEMS_STATE,

  LAYOUT_PANEL_PUSH_MESSAGE,
  LAYOUT_PANEL_REMOVE_MESSAGE,
  
  LAYOUT_CLOSE_MODAL,
  LAYOUT_CLOSE_MODALS,
  LAYOUT_OPEN_MODAL

} from '$actionTypes';

let initialState = {
  modal: {
    modals: []
  },
  navPanel: {
    visible: true,
    messages: [],
    items: [ [{ type: 'cart'}] ],
    loading: false
  }
};

export default (state = initialState, {type, payload}) => {
  let prevState = {...state};
  switch (type) {
    case LAYOUT_PANEL_PUSH_MESSAGE:
      prevState.navPanel.messages.push(payload);
      break;
    
    case LAYOUT_PANEL_REMOVE_MESSAGE:
      prevState.navPanel.messages.shift();
      break;

    case LAYOUT_PANEL_LOADING_TRUE:
      prevState.navPanel.loading = true;
      break;

    case LAYOUT_PANEL_LOADING_FALSE:
      prevState.navPanel.loading = false;
      break;
    
    case LAYOUT_PANEL_SET_ITEMS_STATE:
      prevState.navPanel.items = [payload];
      break;

    case LAYOUT_PANEL_PUSH_ITEMS_STATE:
      prevState.navPanel.items.push(payload);
      break;
    
    case LAYOUT_PANEL_POP_ITEMS_STATE:
      prevState.navPanel.items.pop();
      break;

    case LAYOUT_CLOSE_MODAL:
      prevState.modal.modals.pop()
      break;

    case LAYOUT_CLOSE_MODALS:
      prevState.modal.modals = [];
      prevState.navPanel.items = [prevState.navPanel.items[0]];
      break;
    
    case LAYOUT_OPEN_MODAL:
      prevState.modal.modals.push(payload);
      break;
    
    default:
      return state;
  }

  return prevState;
}
