// $reducers/phonesPage.js
import {
  FETCH_PRODUCTS_SUCCESS,
  FETCH_PRODUCTS_START,
  FETCH_PRODUCTS_FAIL,

  PRODUCTS_FILTER,

  PRODUCTS_SAVE_SCROLL_POS,

  PRODUCTS_IMAGE_LOADED,

  CLEAN_PRODUCTS
} from '$actionTypes';

import { uniqueProducts } from '$helpers';

const initialState = {
  query: '',
  scrollPos: 0,
  products: {},
  ids: [],
  hasMore: true,
  fetching: false,
  error: null
};

export default (state = initialState, {type, payload}) => {
  let prevState = {...state};
  switch (type) {
    case CLEAN_PRODUCTS:
      prevState.ids = [];
      break;

    case FETCH_PRODUCTS_START:
      prevState.fetching = true;
      prevState.error = null;
      break;
      
    case FETCH_PRODUCTS_SUCCESS:
      prevState.products = uniqueProducts(prevState.products, payload);
      prevState.ids = [...prevState.ids, ...payload.map(item => item.id)];
      prevState.fetching = false;
      break;
    
    case PRODUCTS_IMAGE_LOADED:
      prevState.products[payload.id].imageLoaded = true;
      break;
    
    case FETCH_PRODUCTS_FAIL:
      prevState.error = payload;
      break;

    case PRODUCTS_SAVE_SCROLL_POS:
      prevState.scrollPos = payload;
      break;

    case PRODUCTS_FILTER:
      prevState.query = payload;
      break;
    
    default:
      return state
  }

  return prevState;
}
